cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -1011,6 +1012,118 @@ struct GroupedSumFactory {
   InputType argument_type;
 };
 
+// ----------------------------------------------------------------------
+// Product implementation
+
+using arrow::internal::MultiplyWithOverflow;
+
+template <typename Type>
+struct GroupedProductImpl final : public GroupedAggregator {
+  using AccType = typename FindAccumulatorType<Type>::Type;
+  using ProductType = typename TypeTraits<AccType>::CType;
+
+  Status Init(ExecContext* ctx, const FunctionOptions*) override {
+    pool_ = ctx->memory_pool();
+    products_ = TypedBufferBuilder<ProductType>(pool_);
+    valid_ = TypedBufferBuilder<bool>(pool_);
+    out_type_ = TypeTraits<AccType>::type_singleton();
+    return Status::OK();
+  }
```

Comment:
The options are not taken into account (e.g. `min_count = 2`)?

Reply:
Ah, whoops, let me fix that.

Reply:
The hash mean/sum don't support this either so I'll modify all three.